OpenSim/Moco/MocoGoal/MocoImpulseTrackingGoal.cpp

```diff
@@ -0,0 +1,362 @@
+            ++ic) {
+            const auto& path = group.get_contact_force_paths(ic);
+            const auto& contactForce =
+                model.getComponent<SmoothSphereHalfSpaceForce>(path);
+
+            int recordOffset = findRecordOffset(group, contactForce,
+                extForce.get_applied_to_body());
+
+            groupInfo.contacts.push_back(
+                std::make_pair(&contactForce, recordOffset));
+        }
+
+        // Gather the relevant data splines for this contact group.
+        // We assume that the "x", "y", and "z" columns could have been in any
+        // order.
+        const std::string& forceID = extForce.get_force_identifier();
+        groupInfo.refSplines.cloneAndAppend(allRefSplines.get(forceID + "x"));
+        groupInfo.refSplines.cloneAndAppend(allRefSplines.get(forceID + "y"));
+        groupInfo.refSplines.cloneAndAppend(allRefSplines.get(forceID + "z"));
+
+        // Check which frame the contact force data is expressed in.
+        groupInfo.refExpressedInFrame = nullptr;
+        if (extForce.get_force_expressed_in_body() != "ground") {
+            const auto& forceExpressedInBody =
+                extForce.get_force_expressed_in_body();
+            if (model.hasComponent<PhysicalFrame>(forceExpressedInBody)) {
+                groupInfo.refExpressedInFrame =
+                    &model.getComponent<PhysicalFrame>(
+                        forceExpressedInBody);
+            }
+            else if (model.hasComponent<PhysicalFrame>(
+                "./bodyset/" + forceExpressedInBody)) {
+                groupInfo.refExpressedInFrame =
+                    &model.getComponent<PhysicalFrame>(
+                        "./bodyset/" + forceExpressedInBody);
+            }
+            else {
+                OPENSIM_THROW_FRMOBJ(Exception,
+                    "Could not find '{}' in the model or the BodySet.",
+                    forceExpressedInBody);
+            }
+        }
+
+        m_groups.push_back(groupInfo);
+
+        // Check to see if the model contains a MocoScaleFactor associated with
+        // this contact group.
+        std::array<RefPtrMSF, 3> theseScaleFactorRefs;
+        for (int j = 0; j < 3; ++j) {
+            bool foundScaleFactor = false;
+            std::pair<std::string, int> key(group.get_external_force_name(), j);
+            for (const auto& scaleFactor : scaleFactors) {
+                if (m_scaleFactorMap[key] == scaleFactor.getName()) {
+                    theseScaleFactorRefs[j] = &scaleFactor;
+                    foundScaleFactor = true;
+                }
+            }
+            // If we didn't find a MocoScaleFactor for this force direction, set
+            // the reference pointer to null.
+            if (!foundScaleFactor) {
+                theseScaleFactorRefs[j] = nullptr;
+            }
+        }
+        m_scaleFactorRefs.push_back(theseScaleFactorRefs);
```

Comment:
This code block can be updated as follows to reflect the other commented changes for the `MocoScaleFactor`: 

```
// Check to see if the model contains a MocoScaleFactor associated with
// this contact group.
RefPtrMSF thisScaleFactorRef;
bool foundScaleFactor = false;
std::pair<std::string, int> key(group.get_external_force_name(), j);
for (const auto& scaleFactor : scaleFactors) {
    if (m_scaleFactorMap[key] == scaleFactor.getName()) {
        thisScaleFactorRef= &scaleFactor;
        foundScaleFactor = true;
    }
}
// If we didn't find a MocoScaleFactor for this impulse, set
// the reference pointer to null.
if (!foundScaleFactor) {
    thisScaleFactorRef[j] = nullptr;
}
m_scaleFactorRefs.push_back(thisScaleFactorRef);
```